### ISSUE TITLE
mbedtls_ecdsa_read_signature() does not return MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH when expected

### ISSUE THREAD

**Issue by mpg** [Mbed-TLS/mbedtls#2215]
### Description
- Type: Bug
- Priority: Major (affects testing of PSA integration)

## Bug

**OS**  
all

**mbed TLS build:**  
Version: 2.14.0 - but has been there for a long time  
Configuration: default
Compiler and options (if you used a pre-built binary, please indicate how you obtained it):  any

**Expected behavior**   

When passing in a valid signature, but with a `slen` parameter larger than the size of the actual signature, the function should return `MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH` according to its documentation:

```
 * \return          #MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH if there is a valid
 *                  signature in \p sig, but its length is less than \p siglen.
```

**Actual behavior**  

The function returns `MBEDTLS_ERR_ECP_BAD_INPUT_DATA + MBEDTLS_ERR_ASN1_LENGTH_MISMATCH`. ~~In addition to being different from what the documentation says, this error code also violates the conventions by adding two low-level codes, making it unparsable by `mbedtls_strerror()`.~~ ([nope](https://github.com/ARMmbed/mbedtls/issues/2215#issuecomment-440362621))

**Steps to reproduce**  

Apply the following patch then try running the test suite and observe the failure:

```
diff --git a/tests/suites/test_suite_pk.function b/tests/suites/test_suite_pk.function
index 4813f71f7cf7..df3a0a4f79f5 100644
--- a/tests/suites/test_suite_pk.function
+++ b/tests/suites/test_suite_pk.function
@@ -391,6 +391,10 @@ void pk_sign_verify( int type, int sign_ret, int verify_ret )
 
     if( verify_ret == 0 )
     {
+        TEST_ASSERT( mbedtls_pk_verify( &pk, MBEDTLS_MD_SHA256,
+                                        hash, sizeof hash, sig, sig_len + 1 )
+                     == MBEDTLS_ERR_PK_SIG_LEN_MISMATCH );
+
         hash[0]++;
         TEST_ASSERT( mbedtls_pk_verify( &pk, MBEDTLS_MD_SHA256,
                                 hash, sizeof hash, sig, sig_len ) != 0 );
@@ -400,6 +404,7 @@ void pk_sign_verify( int type, int sign_ret, int verify_ret )
         TEST_ASSERT( mbedtls_pk_verify( &pk, MBEDTLS_MD_SHA256,
                                 hash, sizeof hash, sig, sig_len ) != 0 );
         sig[0]--;
+
     }
 
     TEST_ASSERT( mbedtls_pk_sign( &pk, MBEDTLS_MD_SHA256, hash, sizeof hash,
```

**Comment by gilles-peskine-arm** [comment]
> The function returns `MBEDTLS_ERR_ECP_BAD_INPUT_DATA` + `MBEDTLS_ERR_ASN1_LENGTH_MISMATCH`. In addition to being different from what the documentation says, this error code also violates the conventions by adding two low-level codes, making it unparsable by mbedtls_strerror().

ECP uses high-level error codes, so that part is ok.

**Comment by gilles-peskine-arm** [comment]
The treatment of `MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH` in ECDSA has not changed ever since the commit that added it, which goes back to PolarSSL 1.3.6, a little over four years ago. I think that at this point, anybody who expects `MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH` to have the documented behavior would have complained.

The documented behavior is:

> `MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH` if there is a valid signature in sig but its length is less than siglen,

The actual error code in this case is `MBEDTLS_ERR_ECP_BAD_INPUT_DATA + MBEDTLS_ERR_ASN1_LENGTH_MISMATCH`. The function returns `MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH` in a different case, namely, if the buffer contains an ASN.1 sequence with nothing else afterwards, and the ASN.1 sequence contains r, s and some trailing data.

I don't think this special case where the function actually returns `MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH` is useful at all. Given that nobody is asking for it, I don't think we should implement the documented case either. I propose that instead, **we remove `MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH` from both the code and the documentation. `MBEDTLS_ERR_RSA_SIG_LEN_MISMATCH` and therefore `MBEDTLS_ERR_PK_SIG_LEN_MISMATCH` would stay, of course (`MBEDTLS_ERR_PK_SIG_LEN_MISMATCH` only for the sake of RSA).** @mpg @Patater Opinions?

Obviously, whatever we do, we need to test it. I've started to adapt `ecdsa_write_read_random` (https://github.com/gilles-peskine-arm/mbedtls/tree/ecp_sig_len_mismatch-2.14).

**Comment by mpg** [comment]
> > The function returns `MBEDTLS_ERR_ECP_BAD_INPUT_DATA` + `MBEDTLS_ERR_ASN1_LENGTH_MISMATCH`. In addition to being different from what the documentation says, this error code also violates the conventions by adding two low-level codes, making it unparsable by mbedtls_strerror().
> 
> ECP uses high-level error codes, so that part is ok.

My mistake. I made a typo when invoking `programs/utils/strerror` and then drew conclusions too quickly from the unexpected result. Thanks for noticing.

**Comment by mpg** [comment]
@gilles-peskine-arm Thanks for checking the history! I checked out PolarSSL 1.3.6 (the release that introduced this) and added a test for the expected behaviour, which failed as described here. So this has been broken from the start. This is very embarrassing and I hope (and think) we got a bit better since then.

Given that I remember this was asked by a user, I don't understand how they could not notice. Maybe they just realised in the meantime they didn't actually need it?

> I don't think this special case where the function actually returns MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH is useful at all

I agree. And it doesn't match the description either, so we really don't want that.

> Given that nobody is asking for it, I don't think we should implement the documented case either. I propose that instead, we remove MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH from both the code and the documentation. MBEDTLS_ERR_RSA_SIG_LEN_MISMATCH and therefore MBEDTLS_ERR_PK_SIG_LEN_MISMATCH would stay, of course (MBEDTLS_ERR_PK_SIG_LEN_MISMATCH only for the sake of RSA)

Note: there is no `MBEDTLS_ERR_RSA_SIG_LEN_MISMATCH`. The RSA API is old enough that the size of the signature isn't even passed as an argument. The check happens in the PK wrapper and a PK error is generated.

I can no longer find the thread (again, our bookkeeping wasn't as good at the time), but I'm almost certain the original request fort that feature was about ECDSA (I think it was in the context of a bitcoin implementation), and that it was added for RSA only for consistency. So I'd be inclined to remove it entirely.

I think this error code is confusing. A team member recently misunderstood it, and IIRC you were quite surprised by it when you first encountered it too. Let's avoid having confusing bits in our API if we can - in in that case since nobody complained in 4 years, I think we can.
